File: leetcode/380. Insert Delete GetRandom O(1).cpp
class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(m.find(val) != m.end()){
            return false;
        }
        v.push_back(val);
        m[val] = v.size() - 1;
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if(m.find(val) == m.end()){
            return false;
        }
        int idx = m[val];
        int num = v[v.size() - 1];
        
        m[num] = idx;
        m.erase(val);

        v[idx] = num;
        v.pop_back();
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        int idx = rand() % v.size();
        return v[idx];
    }
    
private:
    unordered_map<int, int> m;
    vector<int> v;
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: leetcode/380. Insert Delete GetRandom O(1).py
from random import randint

class RandomizedSet:

    def __init__(self):
        self.dic = {}
        self.lst = []

    def insert(self, val: int) -> bool:
        if val in self.dic:
            return False
        else:
            self.dic[val] = len(self.lst)
            self.lst.append(val)
            return True

    def remove(self, val: int) -> bool:
        if val not in self.dic:
            return False
        else:
            idx = self.dic[val]
            swapped_val = self.lst[-1]
            self.dic[swapped_val] = idx
            self.lst[idx] = swapped_val
            
            self.lst.pop()
            del self.dic[val]
            
            return True

    def getRandom(self) -> int:
        idx = randint(0, len(self.lst) - 1)
        return self.lst[idx]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
